addon/utils/query-cache.js

```diff
@@ -0,0 +1,69 @@
+
+const CACHE_DECAY_TIMEOUT = 5000;
+
+export default class QueryCache {
+  constructor() {
+    this._internalCache = Object.create(null);
+    this.decayId = null;
+  }
+
+  add(type, key, value) {
+    if (!key || !value) {
+      return;
+    }
+
+    if (!this._internalCache[type]) {
+      this._internalCache[type] = Object.create(null);
+    }
+
+    this._internalCache[type][key] = value;
+  }
+
+  remove(type, key) {
+    let typeMap = this._internalCache[type];
+
+    if (typeMap && typeMap[key]) {
+      delete typeMap[key];
+    }
+  }
+
+  clear() {
+    this._internalCache = Object.create(null);
+  }
+
+  get(type, key) {
+    if (!this._internalCache[type]) {
+      return;
+    }
+
+    return this._internalCache[type][key];
+  }
+
+  get types() {
+    return Object.keys(this._internalCache);
+  }
+
+  scheduleDecay(delay = CACHE_DECAY_TIMEOUT, onDecay) {
```

Comment:
IMO, a more flexible approach would be to handle decay per cache entry rather than wiping out the entire cache.

A good use case for this is,

Given a page `foo`, there are 3 api alls `a`, `b`, `c`.

I know `a` will always return the same value, and I want it to last as long as possible.
I know `b` gets updated every 5 mins, so I want to wipe only `b` every 5 mins

`queryRecord` passes in an option, each query should assign a meta property per cache object for the decay, and wipe only the ones that are decayed